QA Process report for v0.37.x (and baseline for v0.34.x) (#9499)
tendermint/tendermint: docs/qa/method.md

_200 Node Test_, and _Rotating Nodes Test_.

[releases]: https://github.com/tendermint/tendermint/blob/v0.37.x/RELEASES.md#large-scale-testnets

# 200 Node Testnet

## Running the test

This section explains how the tests were carried out for reproducibility purposes.

### Infrastructure Requirements

* An account at Digital Ocean (DO), with a high droplet limit (>202)
* The machine to orchestrate the tests should have the following installed:
    * A clone of the [testnet repository][testnet-repo]
        * This repository contains all the scripts mentioned in the reminder of this section
    * [Digital Ocean CLI][doctl]
    * [Terraform CLI][Terraform]
    * [Ansible CLI][Ansible]

### Steps

1. Follow steps 1-4 of the `README.md` at the top of the testnet repository to configure Terraform, and `doctl`.
2. Copy file `testnet200.toml` onto `testnet200.toml` (do NOT commit this change)
3. Follow steps 5-10 of the `README.md` to configure and start the 200 node testnet
    * WARNING: Do NOT forget to run `make terraform-destroy` as soon as you are done with the tests
4. `ssh` into the `testnet-load-runner` and run script XXXXX
    * This will take about 40 mins to run

>     * It is running 90-seconds-long experiments in a loop with different loads

Reviewer: I was thinking to add a comment on verifying whether the blocks are being created and Tendermint is running (as we did when trying to set up the testnet). But I guess this will be apparent from step 5?

Author: No, you got a point. Let me add an extra step

---
order: 1
title: Method
---

> The (first iteration of the) QA process as described [here][releases] is applied to version v0.34.x

Reviewer 2: ```suggestion
The (first iteration of the) QA process as described [in the RELEASES.md document][releases] is applied to version v0.34.x
```

in order to have a set of results acting as benchmarking baseline.
This baseline is then compared with results obtained in later versions.

> Out of the testnet-based test cases described in [releases][releases] we are focusing on two of them:

Reviewer 2: ```suggestion
Out of the testnet-based test cases described in [the releases document][releases] we are focusing on two of them:
```

_200 Node Test_, and _Rotating Nodes Test_.

[releases]: https://github.com/tendermint/tendermint/blob/v0.37.x/RELEASES.md#large-scale-testnets

# Software Dependencies

## Infrastructure Requirements to Run the Tests

* An account at Digital Ocean (DO), with a high droplet limit (>202)
* The machine to orchestrate the tests should have the following installed:
    * A clone of the [testnet repository][testnet-repo]
        * This repository contains all the scripts mentioned in the reminder of this section
    * [Digital Ocean CLI][doctl]
    * [Terraform CLI][Terraform]
    * [Ansible CLI][Ansible]

[testnet-repo]: https://github.com/interchainio/tendermint-testnet
[Ansible]: https://docs.ansible.com/ansible/latest/index.html
[Terraform]: https://www.terraform.io/docs

> [doctl]: https://docs.digitalocean.com/reference/doctl/how-to/install/

Reviewer 2: That testnet repo is currently private, I'm fine making it public, do you think that's the right course of action?

Author: Let's leave it as it is ATM (i.e., private). This is here for reproducibility purposes, people can always ask access.
When we decide what to do with the repo (make it public?, move it inside Tendermeint repo?, etc) then we can come back here and adapt the link.

... for the time being we can add a comment here like: "private repo, ask Tendermint Core Team for access if you are interested". What do you think?

## Requirements for Result Extraction

> * Matlab or Octave
> * Prometheus server installed
> * blockstore DB of one of the full nodes in the testnet
> * Prometheus DB

Reviewer 2: ```suggestion
* Matlab or Octave
* [Prometheus][prometheus] server installed
* blockstore DB of one of the full nodes in the testnet
* Prometheus DB

[prometheus]: https://prometheus.io/
```

Link to prometheus maybe?

# 200 Node Testnet

## Running the test

This section explains how the tests were carried out for reproducibility purposes.

> 1. [If you haven't done it before]

Reviewer 2: This is a very low level of detail, I'm wondering what utility you think there is from describing things in this much detail? Is this for (1) us/future runners of the benchmark or is it for (2) users of the software who want to understand how the benchmarks were produced? 

For (2) I think that this is much too low level and makes it a bit difficult to understand the basic things a user might want to know like "what was the power of the machines?" and "what was the network connectivity and configuration like?" etc that they may want to get at a glance.

Author: When I wrote it, I was rather thinking of (1). Basically improving the bus factor, in case Thane you and I were offline for a while and someone else in the team had to take over these activities.
I have to say that, without this level of detail, I imagine myself trying to do this for the first time (and not being able to Zoom you :-D) and I would be in for a hard time.

Author: Sure thing!

Reviewer 2: Much clearer!

   Follow steps 1-4 of the `README.md` at the top of the testnet repository to configure Terraform, and `doctl`.
2. Copy file `testnet200.toml` onto `testnet.toml` (do NOT commit this change)
3. Set variable `VERSION_TAG` to the git hash that is to be tested.
4. Follow steps 5-10 of the `README.md` to configure and start the 200 node testnet
    * WARNING: Do NOT forget to run `make terraform-destroy` as soon as you are done with the tests
5. As a sanity check, connect to the Prometheus node's web interface and check the graph for the `tendermint_consensus_height` metric.
   All nodes should be increasing their heights.

> 6. `ssh` into the `testnet-load-runner` and run script XXXXX

Reviewer 2: I'll commit this script.

Author: Thanks!

## Software Dependencies

### Infrastructure Requirements to Run the Tests

* An account at Digital Ocean (DO), with a high droplet limit (>202)
* The machine to orchestrate the tests should have the following installed:
    * A clone of the [testnet repository][testnet-repo]
        * This repository contains all the scripts mentioned in the reminder of this section
    * [Digital Ocean CLI][doctl]
    * [Terraform CLI][Terraform]
    * [Ansible CLI][Ansible]

[testnet-repo]: https://github.com/interchainio/tendermint-testnet
[Ansible]: https://docs.ansible.com/ansible/latest/index.html
[Terraform]: https://www.terraform.io/docs
[doctl]: https://docs.digitalocean.com/reference/doctl/how-to/install/

### Requirements for Result Extraction

* Matlab or Octave
* [Prometheus][prometheus] server installed
* blockstore DB of one of the full nodes in the testnet
* Prometheus DB

[prometheus]: https://prometheus.io/

## 200 Node Testnet

### Running the test

This section explains how the tests were carried out for reproducibility purposes.

1. [If you haven't done it before]
   Follow steps 1-4 of the `README.md` at the top of the testnet repository to configure Terraform, and `doctl`.

> 2. Copy file `testnet200.toml` onto `testnet.toml` (do NOT commit this change)

Author: Good point. Fixed

> 3. Set variable `VERSION_TAG` to the git hash that is to be tested.

Author: I reworded it

4. Follow steps 5-10 of the `README.md` to configure and start the 200 node testnet
    * WARNING: Do NOT forget to run `make terraform-destroy` as soon as you are done with the tests
5. As a sanity check, connect to the Prometheus node's web interface and check the graph for the `tendermint_consensus_height` metric.
   All nodes should be increasing their heights.

> 6. `ssh` into the `testnet-load-runner`, then copy script `script/200-node-loadscript.sh` and run it from the load runner node.

Author: Good point. Added

    * This will take about 40 mins to run
    * It is running 90-seconds-long experiments in a loop with different loads
7. Run `make retrieve-data` to gather all relevant data from the testnet into the orchestrating machine
8. Verify that the data was collected without errors
    * at least one blockstore DB for a Tendermint validator
    * the Prometheus database from the Prometheus node
    * for extra care, you can run `zip -T` on the `prometheus.zip` file and (one of) the `blockstore.db.zip` file(s)

> 9. **Run `make terraform-destroy`**

Author: Yes!

3. Set variable `VERSION_TAG` to the git hash that is to be tested.
4. Run `make terraform-apply EPHEMERAL_SIZE=25`
    * WARNING: Do NOT forget to run `make terraform-destroy` as soon as you are done with the tests
5. Follow steps 6-10 of the `README.md` to configure and start the "stable" part of the rotating node testnet
6. As a sanity check, connect to the Prometheus node's web interface and check the graph for the `tendermint_consensus_height` metric.
   All nodes should be increasing their heights.
7. On a different shell,
    * run `make runload ROTATE_CONNECTIONS=X ROTATE_TX_RATE=Y`
    * `X` and `Y` should reflect a load below the saturation point (see, e.g.,
      [this paragraph](./v034/README.md#finding-the-saturation-point) for further info)
8. Run `make rotate` to start the script that creates the ephemeral nodes, and kills them when they are caught up.
    * WARNING: If you run this command from your laptop, the laptop needs to be up and connected for full length
      of the experiment.
9. When the height of the chain reaches 3000, stop the `make rotate` script
10. When the rotate script has made two iterations (i.e., all ephemeral nodes have caught up twice)

>     after height 3000 was reached, stop `make rotate`

Author: Yes, that's what I ended up doing. On my setup, the rotate script had to be _babysitted_ so I couldn't find two slots of 4 hours (the initial length we had agreed) to run (and look after) them.
Further, in the time we ran both tests (v0.34 and v0.37), there were around 30 iterations of the script's loop. I thought this is enough to ascertain that blocksync is working.
Once we move the orchestrating scripts to DO in v0.38.x, we can make this test last longer if we feel it adds value.
